### PR TITLE
Rewrite Dockerfile so build completes

### PR DESCRIPTION
**Pull request by biosafetylvl5**
# Summary
* Fixed Docker build failure by rewriting the Dockerfile
* The Docker build now completes successfully
* This Dockerfile installs rasterio dependencies (`gdal-bin`, `libgdal-dev`).
* The Dockerfile is currently single-staged, but should be converted to multi-stage with volumes

* [x] Required ***existing tests*** pass (ie full_test.sh, others as appropriate)
* [x] Required ***release notes*** added for bugfix
* [x] NO REQUIRED ***unit tests*** added and pass for new/modified functionality
* [x] NO REQUIRED ***integration tests*** (no known tests for the Dockerfile)
* [x] NO REQUIRED ***documentation*** (no known docs for the Dockerfile)
* [x] NO REQUIRED ***updates to other repos*** (no known external repo dependancy)

# Related Issues
fixes NRLMMD-GEOIPS/geoips#449
# Testing Instructions
- Clone geoips
- Run `docker build -t geoips .`
- Confirm the Docker build completes successfully without hanging on installing rasterio
- Run `bash` and execute integration and unit tests.

# Testing Output
- Command line output confirms successful Docker build with the updated Dockerfile
``` bash
gwynu@Gwyns-MacBook-Pro geoips % docker build -t geoips .          
[+] Building 341.8s (20/20) FINISHED                                                                                                                                                                             docker:desktop-linux
 => [internal] load build definition from Dockerfile                                                                                                                                                                             0.0s
 => => transferring dockerfile: 1.71kB                                                                                                                                                                                           0.0s
 => [internal] load metadata for docker.io/library/python:3.10-slim-bullseye                                                                                                                                                     1.2s
 => [internal] load .dockerignore                                                                                                                                                                                                0.0s
 => => transferring context: 72B                                                                                                                                                                                                 0.0s
 => [ 1/15] FROM docker.io/library/python:3.10-slim-bullseye@sha256:e89739e9b11753ace337116c9086c4ba780c0c3daff118a7c680783cf7afada1                                                                                             0.0s
 => [internal] load build context                                                                                                                                                                                                0.1s
 => => transferring context: 80.23kB                                                                                                                                                                                             0.0s
 => CACHED [ 2/15] RUN apt-get update && apt-get -y upgrade                                                                                                                                                                      0.0s
 => [ 3/15] RUN apt-get install -y wget git libopenblas-dev imagemagick g++ make imagemagick                                                                                                                                    53.4s
 => [ 4/15] RUN apt-get update && apt-get install -y software-properties-common                                                                                                                                                 10.8s 
 => [ 5/15] RUN add-apt-repository -y ppa:ubuntugis/ppa                                                                                                                                                                          2.6s 
 => [ 6/15] RUN apt-get install -y gdal-bin libgdal-dev                                                                                                                                                                         29.5s 
 => [ 7/15] RUN pip install -U pip                                                                                                                                                                                               2.7s 
 => [ 8/15] RUN pip install rasterio                                                                                                                                                                                            73.0s 
 => [ 9/15] WORKDIR /app/geoips_packages                                                                                                                                                                                         0.0s 
 => [10/15] RUN getent group 25000 > /dev/null || groupadd -g 25000 geoips_user     && useradd --gid 25000 --uid 25000 -l -m geoips_user                                                                                         0.2s 
 => [11/15] RUN mkdir -p /output /data /data                                                                                                                                                                                     0.3s 
 => [12/15] RUN chown 25000:25000 /output /data /data                                                                                                                                                                            0.3s 
 => [13/15] COPY --chown=25000:25000 . /app/geoips_packages/geoips                                                                                                                                                               0.3s 
 => [14/15] WORKDIR /app/geoips_packages/geoips                                                                                                                                                                                  0.0s 
 => [15/15] RUN cd /app/geoips_packages/geoips     && pip install --no-cache . #".[doc,lint,test,debug]"                                                                                                                       166.3s
 => exporting to image                                                                                                                                                                                                           1.3s 
 => => exporting layers                                                                                                                                                                                                          1.3s 
 => => writing image sha256:f779eb86a782aa2d0cd7933b7403d75a82eed8569400038bc06b24be1a15d797                                                                                                                                     0.0s 
 => => naming to docker.io/library/geoips 
```
- Run `bash` and execute integration and unit tests.
```
gwynu@Gwyns-MacBook-Pro geoips % docker run -ti --entrypoint=bash geoips

geoips_user@b4de8ecf935e:/app/geoips_packages/geoips$ whoami

geoips_user
```
``` bash
geoips_user@b4de8ecf935e:/app/geoips_packages/geoips$ $GEOIPS_PACKAGES_DIR/geoips/tests/integration_tests/base_install.sh

  

Install log: /output/logs/20240315.003635_install.log

  

Install log: /output/logs/20240315.003635_install.log

SUCCESS: 'imagemagick' appears to be installed successfully

    /usr/bin/compare

  

Install log: /output/logs/20240315.003635_install.log

SUCCESS: 'git' appears to be installed successfully

    /usr/bin/git

  

Install log: /output/logs/20240315.003635_install.log

SUCCESS: 'python' appears to be installed successfully

    /usr/local/bin/python

  

Install log: /output/logs/20240315.003635_install.log

Installing tests data repo test_data_amsr2 .... 

  /data/test_data_amsr2/ from https://io.cira.colostate.edu/s/FmWwX2ft7KDQ8N9/download/test_data_amsr2.tgz

SUCCESS: Pulled and decompressed test_data_amsr2 from https://io.cira.colostate.edu/s/FmWwX2ft7KDQ8N9/download/test_data_amsr2.tgz

  

Install log: /output/logs/20240315.003856_install.log

Installing tests data repo test_data_noaa_aws .... 

  /data/test_data_noaa_aws/ from https://io.cira.colostate.edu/s/fkiPS3jyrQGqgPN/download/test_data_noaa_aws.tgz

SUCCESS: Pulled and decompressed test_data_noaa_aws from https://io.cira.colostate.edu/s/fkiPS3jyrQGqgPN/download/test_data_noaa_aws.tgz
```
``` bash
geoips_user@b4de8ecf935e:/app/geoips_packages/geoips$ create_plugin_registries

15_005128 create_plugin_registries.py:791  INTERACTIVE: Skipping module 'tc_file_naming' from 'geoips', interface_name is 'None'

15_005128 create_plugin_registries.py:791  INTERACTIVE: Skipping module 'remss_reader' from 'geoips', interface_name is 'None'

15_005128 create_plugin_registries.py:791  INTERACTIVE: Skipping module 'geostationary_geolocation' from 'geoips', interface_name is 'None'

15_005128 create_plugin_registries.py:791  INTERACTIVE: Skipping module 'hrit_reader' from 'geoips', interface_name is 'None'

15_005128 create_plugin_registries.py:791  INTERACTIVE: Skipping module 'interp_pyresample' from 'geoips', interface_name is 'None'

15_005128 create_plugin_registries.py:791  INTERACTIVE: Skipping module 'interp_scipy' from 'geoips', interface_name is 'None'

15_005128 create_plugin_registries.py:791  INTERACTIVE: Skipping module 'boxdefinitions' from 'geoips', interface_name is 'None'

15_005128 create_plugin_registries.py:295  INTERACTIVE: Writing /home/geoips_user/.local/lib/python3.10/site-packages/geoips/registered_plugins.json
```
``` bash
geoips_user@b4de8ecf935e:/app/geoips_packages/geoips$ $GEOIPS_PACKAGES_DIR/geoips/tests/integration_tests/base_test.sh

  

Install log: /output/logs/20240315.005148_install.log

  

Install log: /output/logs/20240315.005148_install.log

SUCCESS: 'imagemagick' appears to be installed successfully

    /usr/bin/compare

  

Install log: /output/logs/20240315.005148_install.log

SUCCESS: 'git' appears to be installed successfully

    /usr/bin/git

  

Install log: /output/logs/20240315.005148_install.log

SUCCESS: 'python' appears to be installed successfully

    /usr/local/bin/python

  

Install log: /output/logs/20240315.005148_install.log

SUCCESS: tests data repo test_data_amsr2 appears to be installed successfully

    drwxr-xr-x 5 geoips_user geoips_user 4096 Apr 4 2023 /data/test_data_amsr2

  

Install log: /output/logs/20240315.005148_install.log

SUCCESS: tests data repo test_data_noaa_aws appears to be installed successfully

    drwxr-xr-x 3 geoips_user geoips_user 4096 Apr 4 2023 /data/test_data_noaa_aws

  

Fri Mar 15 00:51:48 UTC 2024 Running pre, log: /output/logs/20240315/20240315.005148_geoips_base/test_all_geoips_base.log

  

Fri Mar 15 00:51:48 UTC 2024  Running amsr2.config_based_no_compare.sh

/app/geoips_packages/geoips/tests/scripts/amsr2.config_based_no_compare.sh

/output/logs/20240315/20240315.005148_geoips_base/test_all_geoips_base.log_amsr2.config_based_no_compare.sh.log

        Return: 0

  

Fri Mar 15 00:52:42 UTC 2024  Running amsr2_ocean.tc.windspeed.imagery_clean.sh

/app/geoips_packages/geoips/tests/scripts/amsr2_ocean.tc.windspeed.imagery_clean.sh

/output/logs/20240315/20240315.005148_geoips_base/test_all_geoips_base.log_amsr2_ocean.tc.windspeed.imagery_clean.sh.log

        Return: 0

  

Fri Mar 15 00:52:47 UTC 2024  Running test_interfaces

test_interfaces

/output/logs/20240315/20240315.005148_geoips_base/test_all_geoips_base.log_test_interfaces.log

        Return: 0

  

  

Fri Mar 15 00:53:01 UTC 2024  Running post, final results in /output/logs/20240315/20240315.005148_geoips_base/test_all_geoips_base.log

  

Package: geoips_base

Total run time: 73 seconds

Number data types run: 3

Number data types failed: 0

Fri Mar 15 00:53:01 UTC 2024
```

### CODE REVIEW

**Comment by biosafetylvl5** [comment]
To install the `rasterio` you need `g++` and for `pypublicdecompwt` you need `make`. Should we add `make` to the [expert installation instructions](https://github.com/NRLMMD-GEOIPS/geoips/blob/9aafb0d1712eedcbbf551b6b069d40af484bd7ca/docs/source/starter/expert_installation.rst#id5)?

**Comment by mindyls** [comment]
> To install the `rasterio` you need `g++` and for `pypublicdecompwt` you need `make`. Should we add `make` to the [expert installation instructions](https://github.com/NRLMMD-GEOIPS/geoips/blob/9aafb0d1712eedcbbf551b6b069d40af484bd7ca/docs/source/starter/expert_installation.rst#id5)?

I think so !

File: Dockerfile [hunk 1]
```diff
@@ -1,72 +1,22 @@
...
-# If not provided, images will be pulled from Docker Hub
-ARG EXT_REGISTRY=""
-ARG IMAGE_TAG="11.4"
+FROM python:3.10-slim-bullseye
 
-FROM ${EXT_REGISTRY}docker.io/library/debian:${IMAGE_TAG} as proj
-# Renew build args for the stage
-ARG EXT_REGISTRY IMAGE_TAG
+RUN apt-get update && apt-get -y upgrade
+RUN apt-get install -y wget git libopenblas-dev imagemagick g++ make imagemagick
```
Comment (mindyls) [inline Dockerfile:4]: Should be able to remove imagemagick here.

File: Dockerfile [hunk 2]
```diff
@@ -1,72 +1,22 @@
...
 
-ARG PROJ_VERSION='9.0.1'
+RUN apt-get update && apt-get install -y software-properties-common
+RUN add-apt-repository -y ppa:ubuntugis/ppa
+RUN apt-get install -y gdal-bin libgdal-dev
+RUN pip install -U pip 
 
-WORKDIR /build
+# could install the rest of geoips dependancies here
+RUN pip install rasterio
```
Comment (mindyls) [inline Dockerfile:12]: Does this need to be installed first? It does not successfully build during the geoips installation?  It should be included as a dependency in geoips?

Comment (biosafetylvl5) [reply Dockerfile:12]: It doesn't need to be installed first, but if the end goal is a super-small image I'm thinking it'd be advantageous to build geoips dependencies and then pull them over to a new image without their build tooling

Comment (mindyls) [reply Dockerfile:12]: Ahh, so this is more of a proof of concept / start of a multistage build.  Got it.

File: Dockerfile (hunk 1 above)
Comment (biosafetylvl5) [reply Dockerfile:4]: I figured that'd be best done by merging with main (now that PR #340 has been merged) for organizations sake - would it be more conventional to remove in this PR?

Comment (mindyls) [reply Dockerfile:4]: In the interest of making this a minimal/functional Dockerfile, I think it is probably fine to remove it here (basically re-evaluating all the dependencies in this dockerfile independently of that other PR).  Maybe @jsolbrig has a different opinion?

Comment (biosafetylvl5) [reply Dockerfile]: Fixed in most recent commit :)

**Review by jsolbrig — APPROVED** [review]
I made one change to this to call `create_plugin_registries` in the last line of the file. Otherwise, this looks good. I was able to build and run the docker image and could successfully run `tests/integration_tests/base_test.py`.

**Comment by biosafetylvl5** [comment]
Thanks Jeremy! I fixed a typo in your commit. LGTM :)

**Review by mindyls — APPROVED** [review]

**Pull request merged**